R/tfpscanner.R

```diff
@@ -62,698 +70,796 @@ tfpscan <- function(tre,
+    tx <- sort(table(lins), decreasing = TRUE) / length(lins)
+    if (length(tx) > 1) {
+      y <- as.data.frame(tx)
+      colnames(y) <- c("Lineage", "Frequency")
+    } else {
+      y <- data.frame(Lineage = lins[1], Frequency = 1)
+    }
+    y <- y[seq_len(min(nrow(y), maxrows)), ]
+    y$Frequency <- paste0(round(y$Frequency * 100), "%")
+    paste(knitr::kable(y, "simple"), collapse = "\n") # convert to string
+  }
+
+  .region_summary <- function(tips, maxrows = 5) {
+    if (is.null(tips)) {
+      return("")
+    }
+    regs <- amd$region[match(tips, amd$sequence_name)]
+    tx <- sort(table(regs), decreasing = TRUE) / length(regs)
+    if (length(tx) > 1) {
+      y <- as.data.frame(tx)
+      colnames(y) <- c("Region", "Frequency")
+    } else {
+      y <- data.frame(Region = regs[1], Frequency = 1)
+    }
+    y <- y[seq_len(min(nrow(y), maxrows)), ]
+    y$Frequency <- paste0(round(y$Frequency * 100), "%")
+    paste(knitr::kable(y, "simple"), collapse = "\n") # convert to string
+  }
+
+  .cluster_tree <- function(tips) { # tr2, amd # tre?
+    tr <- ape::keep.tip(tre, tips)
+    gtr <- ggtree::ggtree(tr)
+    mutlist <- strsplit(amd[match(tips, amd$sequence_name), ]$mutations, split = "\\|")
+    # ~ sharedmut = Reduce( intersect,  mutlist )
+    mutthresh <- .75
+    tx <- table(do.call(c, mutlist))
+    tx <- tx / ape::Ntip(tr)
+    sharedmut <- names(tx)[tx >= mutthresh] # TODO need a better way to get sharedmut
+    segregatingmut <- lapply(mutlist, function(x) setdiff(x, sharedmut))
+    allsegregating <- Reduce(union, segregatingmut)
+
+    # remove stops from allseg
+    allsegregating <- allsegregating[!grepl(allsegregating,
+      pattern = "[*]$"
+    )]
+    allsegregating <- allsegregating[!grepl(allsegregating,
+      pattern = ":[*]"
+    )]
+    annots <- rep("", ape::Ntip(tr) + ape::Nnode(tr))
+
+    if (length(allsegregating) > 0) {
+      allseg1 <- substr(regmatches(allsegregating, regexpr(allsegregating,
+        pattern = ":[A-Z]"
+      )), 2, 2)
+      allseg2 <- regmatches(allsegregating, regexpr(allsegregating,
+        pattern = "[A-Z*]$"
+      ))
+      sites_post <- regmatches(allsegregating, regexpr(allsegregating,
+        pattern = ":.*$"
+      ))
+      sites_post <- substr(sites_post, 3, nchar(sites_post) - 1)
+      sites_pre <- regmatches(allsegregating, regexpr(allsegregating,
+        pattern = "^.*:"
+      ))
+      sites <- paste0(sites_pre, sites_post)
+
+      aas <- c()
+      aaslist <- lapply(seq_along(allsegregating), function(i) {
+        do.call(rbind, lapply(
+          mutlist,
+          function(x) ifelse(allsegregating[i] %in% x, allseg1[i], allseg2[i])
+        ))
+      })
+      aas <- do.call(cbind, aaslist)
+      colnames(aas) <- allsegregating # sites
+      rownames(aas) <- tr$tip.label
+      aas <- aas[, order(sites)]
+      aas[is.na(aas)] <- "X"
+      sites <- sort(sites)
+      aas2 <- phangorn::as.phyDat(aas, type = "AA")
```

Comment:
there was an `aas1` in JR:master that had been commented out in mrc-ide:master. This has been removed